satellite/repair/repairer/ec.go

```diff
@@ -0,0 +1,468 @@
+
+	for currentLimitIndex, limit := range limits {
+		if limit == nil {
+			continue
+		}
+
+		currentLimitIndex, limit := currentLimitIndex, limit
+		limiter.Go(ctx, func() {
+			cond.L.Lock()
+			defer cond.Signal()
+			defer cond.L.Unlock()
+
+			for {
+				if successfulPieces >= es.RequiredCount() {
+					// already downloaded minimum number of pieces
+					cond.Broadcast()
+					return
+				}
+				if successfulPieces+inProgress+unusedLimits < es.RequiredCount() {
+					// not enough available limits left to get required number of pieces
+					cond.Broadcast()
+					return
+				}
+
+				if successfulPieces+inProgress >= es.RequiredCount() {
+					cond.Wait()
+					continue
+				}
+
+				unusedLimits--
+				inProgress++
+				cond.L.Unlock()
+
+				downloadedPiece, err := ec.downloadAndVerifyPiece(ctx, limit, privateKey, pieceSize, es.ErasureShareSize())
+				cond.L.Lock()
+				inProgress--
+				if err != nil {
+					ec.log.Debug("Failed to download pieces for repair.", zap.Error(err))
+					return
+				}
+
+				pieceReaders[currentLimitIndex] = ioutil.NopCloser(bytes.NewReader(downloadedPiece))
+				successfulPieces++
+
+				return
+			}
+		})
+	}
+
+	limiter.Wait()
+
+	if successfulPieces < es.RequiredCount() {
+		return nil, Error.New("couldn't download enough pieces, number of successful downloaded pieces (%d) is less than required number (%d)", successfulPieces, es.RequiredCount())
+	}
+
+	fec, err := infectious.NewFEC(es.RequiredCount(), es.TotalCount())
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	esScheme := eestream.NewUnsafeRSScheme(fec, es.ErasureShareSize())
+	expectedSize := pieceSize * int64(es.RequiredCount())
+
+	decodeReader := eestream.DecodeReaders(ctx, cancel, ec.log.Named("decode readers"), pieceReaders, esScheme, expectedSize, 0, false)
+
+	return decodeReader, nil
+}
+
+// downloadAndVerifyPiece downloads a piece from a storagenode,
+// expects the original order limit to have the correct piece public key,
+// and expects the hash of the data to match the signed hash provided by the storagenode.
+func (ec *ECRepairer) downloadAndVerifyPiece(ctx context.Context, limit *pb.AddressedOrderLimit, privateKey storj.PiecePrivateKey, pieceSize int64, bufferSize int) (data []byte, err error) {
+	// contact node
+	ps, err := ec.dialPiecestore(ctx, &pb.Node{
+		Id:      limit.GetLimit().StorageNodeId,
+		Address: limit.GetStorageNodeAddress(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	downloader, err := ps.Download(ctx, limit.GetLimit(), privateKey, 0, pieceSize)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = errs.Combine(err, downloader.Close()) }()
+
+	var calculatedHash []byte
+	pieceBytes := make([]byte, 0, pieceSize)
+	buffer := make([]byte, bufferSize)
+	newHash := pkcrypto.NewHash()
+
+	for {
+		// download full piece
+		n, readErr := downloader.Read(buffer)
+		if readErr == io.EOF {
+			calculatedHash = newHash.Sum(buffer[:n])
```

Comment:
We were originally doing something like that, but we thought maybe it would be preferable to hash the piece as it was being read in rather than after `ReadAll`. Now, I don't know if there is any advantage to how we are currently doing it.

Reply:
only thing I can think off is performance and readability. Maybe you can compare if there is a difference between both ways with simple benchmark?

Reply:
```
// Sum appends the current hash to b and returns the resulting slice.
// It does not change the underlying hash state.
    Sum(b []byte) []byte
```
so the `b` here should already be a hash and `Sum` only appends that to the current hash not actually hashing it and then append? Just want to make sure I understand

Reply:
`Sum` just appends `b` to current hash (it doesn't have to be hash also, might be just some bytes) without hash processing and returns `hash + b`. In our code `Sum` has mainly `nil` param because we want just pure hash without additions.